book/scheduling-and-threading.md

```diff
@@ -93,137 +44,130 @@ class Task:
         self.args = args
         self.__name__ = "task"
 
-    def __call__(self):
+    def run(self):
         self.task_code(*self.args)
         self.task_code = None
         self.args = None
 ```
 
-Also define a new `TaskRunner` class to manage the task queues and run them. It
-will have a list of `Task`s, a method to add a `Task`, and a method to run once
-through the event loop. Implement a simple scheduling heuristic in
-`run` that executes one task each time through, if there is one to run.
+The point of a task is that it can be created at one point, and then
+run at some later point by a task runner of some kind.[^event-loop] In
+our browser, the task runner will store tasks in a first-in first-out
+queue:
 
-``` {.python expected=False}
+[^event-loop]: Event loops, which we discussed in [Chapter
+2](graphics.md#eventloop) and [Chapter
+11](visual-effects.md#sdl-creates-the-window) are one kind of task
+runner, where the tasks to run are provided by the operating system.
+
+``` {.python replace=(self)/(self%2c%20tab)}
 class TaskRunner:
     def __init__(self):
         self.tasks = []
 
     def schedule_task(self, callback):
         self.tasks.append(callback)
+```
+
+When the time comes to run a task, our task runner can just remove
+the first task from the queue and run it:
 
+``` {.python expected=False}
+class TaskRunner:
     def run(self):
         if len(self.tasks) > 0):
             task = self.tasks.pop(0)
-            task()
-```
-
-Finally we just need to modify the main event loop to run the task runner each
-time, in case there is a script task to execute:
+            task.run()
 
-``` {.python expected=False}
-if __name__ == "__main__":
-    # ...
-    while True:
-        while sdl2.SDL_PollEvent(ctypes.byref(event)) != 0:
-            # ...
-            browser.tabs[browser.active_tab].task_runner.run()
+class Tab:
+    def __init__(self):
+        self.task_runner = TaskRunner()
 ```
 
-Of course, this is all pointless at the moment, since there aren't any tasks
-yet. Let's try out this infrastructure on a simple example: evaluating scripts
-after load. Currently the browser just evaluates them right away, but instead
-let's make script evaluationt a task.
+This is a pretty simplistic of choosing which task to run next, and
+real browsers have sophisticated *schedulers* which consider things
+like XXX.
```

Comment:
?

Reply:
Ah, yes, I was hoping you could tell me what kinds of things the Chrome scheduler, for example, considers.